### PR TITLE
types/datum.go: casting zero to enum should not change value

### LINKED ISSUE #29525: Alter table incorrectly modified enum internal value.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (id INT AUTO_INCREMENT PRIMARY KEY,c1 ENUM('a', '', 'b'));
SET sql_mode='';
INSERT INTO t1 (c1) VALUES (0);
SELECT id, c1 + 0, c1 FROM t1; 
ALTER TABLE t1 CHANGE c1 c1 ENUM('a', '') NOT NULL;
SELECT id, c1 + 0, c1 FROM t1; 
INSERT INTO t1 (c1) VALUES (0); -- will not match the on disk value as first insert
SELECT id, c1 + 0, c1 FROM t1; 
```
### 2. What did you expect to see? (Required)
```sql
mysql> DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.02 sec)

mysql> CREATE TABLE t1 (id INT AUTO_INCREMENT PRIMARY KEY,c1 ENUM('a', '', 'b'));
Query OK, 0 rows affected (0.02 sec)

mysql> SET sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> INSERT INTO t1 (c1) VALUES (0);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+------+
| id | c1 + 0 | c1   |
+----+--------+------+
|  1 |      0 |      |
+----+--------+------+
1 row in set (0.00 sec)

mysql> ALTER TABLE t1 CHANGE c1 c1 ENUM('a', '') NOT NULL;
Query OK, 1 row affected (0.05 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+----+
| id | c1 + 0 | c1 |
+----+--------+----+
|  1 |      0 |    |
+----+--------+----+
1 row in set (0.00 sec)

mysql> INSERT INTO t1 (c1) VALUES (0); -- will not match the on disk value as first insert
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+----+
| id | c1 + 0 | c1 |
+----+--------+----+
|  1 |      0 |    |
|  2 |      0 |    |
+----+--------+----+
2 rows in set (0.00 sec)

```
### 3. What did you see instead (Required)
```sql
mysql>  DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.30 sec)

mysql> CREATE TABLE t1 (id INT AUTO_INCREMENT PRIMARY KEY, c1 ENUM('a', '', 'b') );
Query OK, 0 rows affected (0.07 sec)

mysql> SET sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> INSERT INTO t1 (c1) VALUES (0);
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+------+
| id | c1 + 0 | c1   |
+----+--------+------+
|  1 |      0 |      |
+----+--------+------+
1 row in set (0.00 sec)

mysql> ALTER TABLE t1 CHANGE c1 c1 ENUM('a', '') NOT NULL;
Query OK, 0 rows affected (2.85 sec)

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+----+
| id | c1 + 0 | c1 |
+----+--------+----+
|  1 |      2 |    |
+----+--------+----+
1 row in set (0.00 sec)

mysql> INSERT INTO t1 (c1) VALUES (0); -- will not match the on disk value as first insert
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> SELECT id, c1 + 0, c1 FROM t1; 
+----+--------+----+
| id | c1 + 0 | c1 |
+----+--------+----+
|  1 |      2 |    |
|  2 |      0 |    |
+----+--------+----+
2 rows in set (0.00 sec)

```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.2.2
Edition: Community
Git Commit Hash: da1c21fd45a4ea5900ac16d2f4a248143f378d18
Git Branch: heads/refs/tags/v5.2.2
UTC Build Time: 2021-10-20 06:03:45
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #29525

Problem Summary:

Casting 0 to enum('') should not change the underlying value.

```sql
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (id INT AUTO_INCREMENT PRIMARY KEY,c1 ENUM('a', '', 'b'));
SET sql_mode='';
INSERT INTO t1 (c1) VALUES (0);
ALTER TABLE t1 CHANGE c1 c1 ENUM('a', '') NOT NULL;  -- It should NOT change 0 to 2
```

### What is changed and how it works?

If the enum value is zero, keep the zero value unchanged regardless of parsing.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix an issue that changing column type modifies the existing enum zero value.
```

### CODE REVIEW

File: executor/executor_test.go [hunk 1]
```diff
@@ -3315,6 +3315,16 @@ func (s *testSuite) TestEmptyEnum(c *C) {
 	tk.MustQuery("select * from t").Check(testkit.Rows("", ""))
 	tk.MustExec("insert into t values (null)")
 	tk.MustQuery("select * from t").Check(testkit.Rows("", "", "<nil>"))
+
+	// Test https://github.com/pingcap/tidb/issues/29525.
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (id int auto_increment primary key, c1 enum('a', '', 'c'));")
+	tk.MustExec("insert into t(c1) values (0);")
+	tk.MustQuery("select id, c1+0, c1 from t;").Check(testkit.Rows("1 0 "))
+	tk.MustExec("alter table t change c1 c1 enum('a', '') not null;")
```
Comment (AilinKid) [inline executor/executor_test.go:3324]: you can change `enum('a', '')` to `enum('a', 'b')`, mysql can do it without any warnings

File: types/datum.go [hunk 2]
```diff
@@ -1500,6 +1500,9 @@ func (d *Datum) convertToMysqlEnum(sc *stmtctx.StatementContext, target *FieldTy
 		e, err = ParseEnum(target.Elems, d.GetString(), target.Collate)
 	case KindMysqlEnum:
 		e, err = ParseEnum(target.Elems, d.GetMysqlEnum().Name, target.Collate)
+		if err == nil && d.i == 0 {
```
Comment (AilinKid) [inline types/datum.go:1503]: actually `value=0 name=""` is a truncated enum (Not valid one), maybe we can just it before parserEnum, and skip parse process

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: types/datum.go [hunk 3]
```diff
@@ -1499,7 +1499,11 @@ func (d *Datum) convertToMysqlEnum(sc *stmtctx.StatementContext, target *FieldTy
 	case KindString, KindBytes, KindBinaryLiteral:
 		e, err = ParseEnum(target.Elems, d.GetString(), target.Collate)
 	case KindMysqlEnum:
-		e, err = ParseEnum(target.Elems, d.GetMysqlEnum().Name, target.Collate)
+		if d.i == 0 {
```
Comment (wjhuang2016) [inline types/datum.go:1502]: Please add a comment for it.

**Review by AilinKid — APPROVED** [review]
LGTM

File: types/datum.go (hunk 3 above)
Comment (tangenta) [reply types/datum.go:1502]: Updated.

**Review by wjhuang2016 — COMMENTED** [review]
If we add a unique index, and use point get plan, we still get the wrong result.
```
alter table t1 add unique index idx(c1);
select * from t1 where c1 = 0;
```

Besides, can we make sure always pass values greater than 0 to ParseEnumValue() 
```
func ParseEnumValue(elems []string, number uint64) (Enum, error) {
	if number == 0 || number > uint64(len(elems)) {
		return Enum{}, errors.Errorf("number %d overflow enum boundary [1, %d]", number, len(elems))
	}

	return Enum{Name: elems[number-1], Value: number}, nil
}
```

I'm afraid it's not a simple problem. If so, we may need a general solution for it.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**